### PR TITLE
Custom CSS, FileBrowser improvements

### PR DESCRIPTION
**Pull request by sharkykh**
Requested in #3457 

Plus changed the FileBrowser so it would be easier to filter other types (extensions) of files in the future.
And changed it so folders would show up before files, like in a regular file browser.

### CODE REVIEW

**Comment by SickRage2** [comment]
Thanks for the pull request! Before a real human comes by, please make sure your PR has all of the below criteria checked

- [x] Give a description on what the PR is for.
- [x] Make sure your PR is based on the DEVELOP branch
- [x] Don't send big changes all at once. Split up big PRs into multiple smaller PRs that are easier to manage and review

Please make sure you also read [contribution guide](https://github.com/SickRage/SickRage/blob/master/contributing.md) and followed all of the steps.

Thanks!

File: gui/slick/js/core.js [hunk 1]
```diff
@@ -443,7 +443,8 @@ var SICKRAGE = {
         },
         index: function() {
             $('#log_dir').fileBrowser({title: 'Select log file folder location'});
-            $('#sickrage_background_path').fileBrowser({title: 'Select Background Image', key: 'sickrage_background_path', includeFiles: 1, imagesOnly: 1});
+            $('#sickrage_background_path').fileBrowser({title: 'Select Background Image', key: 'sickrage_background_path', includeFiles: 1, fileTypes: ['images']});
+            $('#custom_css_path').fileBrowser({title: 'Select CSS file', key: 'sickrage_background_path', includeFiles: 1, fileTypes: ['css']});
```
Comment (miigotu) [inline gui/slick/js/core.js]: Wrong key here?

**Review by miigotu — COMMENTED** [review]
Not sure of the usefulness of the custom css. Will this be just another generally unused feature?  Mostly I try to avoid adding features unless they are expected to be widely appreciated/used, and are sure to not result in any extra issue reports.

In other words, we have lots of broken stuff to fix right now, I'm weary of creating more bugs.

Side note: I think this PR should add a [[shares]] domain to the config, with subdomains we can loop over and add to the file browser like we do in foldersAtPath when adding windows drives.

```config
[shares]
  [[nas]]
    user = miigotu
    password = sickrage
    server = HostNameOrIP
    share = sharedFolderName

  [[pi]]
    user = miigotu
    password = sickrage
    server = piHostNameOrIP
    share = sharedFolderName
```

Then in the foldersAtPath, something like this:

```python
for name, share in six.iteritems(sickbeard.CFG.get('shares')):
    if not (share.get('server') and share.get('share')):
        continue

    loc = ''
    if share.get('user'):
        loc += share.get('user')
        if share.get('password'):
            loc += ':{0}'.format(helpers.decrypt(share.get('password')))
        loc += '@'
    loc += '{0}\\{1}'.format(share.get('server'), share.get('share'))
    loc = '\\\\{0}'.format(loc)
    entries.append({'name': name, 'path': loc})
```

Ofc setting the values from a submitted form uses a simple process:
```python
sickbeard.CFG['shares'][name]['server'] = server
sickbeard.CFG['shares'][name]['share'] = share
sickbeard.CFG['shares'][name]['user'] = user
if password:
    sickbeard.CFG['shares'][name]['password'] = helpers.encrypt(password)
sickbeard.save_config()
```

File: gui/slick/js/core.js (hunk 1 above)
Comment (sharkykh) [reply gui/slick/js/core.js]: Yep. Fixing

File: sickbeard/webserve.py [hunk 2]
```diff
@@ -652,24 +652,32 @@ def sickrage_background(self):
                 return content.read()
         return None
 
+    def custom_css(self):
```
Comment (miigotu) [inline sickbeard/webserve.py:655]: btw, a route exactly like this is all you needed for the locale json lol, in the ui class in webserve, so a call to /ui/locale.json would return the file for GUI_LANG =P

Comment (sharkykh) [reply sickbeard/webserve.py:655]: I'll probably change it when I won't be so tired haha

Comment (sharkykh) [reply sickbeard/webserve.py:655]: english not good so right now
as you can see
:laughing:

**Comment by sharkykh** [comment]
@miigotu  so far, so good.
Just need to adjust some settings like the patch target... maybe we should remove it altogether
`codecov/project = overall coverage change`
`codecov/patch = coverage of changed lines only`

**Comment by sharkykh** [comment]
I'm adding 3 tests for the functions in browser.py so don't merge yet

**Comment by miigotu** [comment]
You didn't answer my question about the point of this or, or input on the feature I'd like to add in this one =p

**Comment by sharkykh** [comment]
Wow I didn't even see that comment. One sec

**Comment by sharkykh** [comment]
> Not sure of the usefulness of the custom css. Will this be just another generally unused feature?  Mostly I try to avoid adding features unless they are expected to be widely appreciated/used, and are sure to not result in any extra issue reports.
> 
> In other words, we have lots of broken stuff to fix right now, I'm weary of creating more bugs.

Well, I'll be honest;
I was following a single user's request, and I just thought that it's a simple addition and I was in the mood, and so **I don't really have a position on the matter.**

> Side note: I think this PR should add a [[shares]] domain to the config, with subdomains we can loop over and add to the file browser like we do in foldersAtPath when adding windows drives.

Sounds very much like you want me to start working on https://github.com/SickRage/SickRage/issues/3359#issuecomment-285451899 😀 
I'll work on that tomorrow. I'm awake since 7am and it's 11:30pm here 😩

**Comment by miigotu** [comment]
I just usually try my best to avoid new/extra config options that most people either won't use/need or understand.
When it comes to features we have a really large list of features, with votes on them to know how much the community wants them on http://feathub.com/SickRage/SickRage

It's not a big deal, we can allow this one because you are right, it's not much to implement it, but in general we should spend time on stuff that affects the highest number of users as possible... in a good way.

The post processing one is annoying but we need to finish that too x.X
As for the windows shares thing, I just think that will help a lot of people.
We could also possibly look through all local network IP's and call this for each one on startup https://github.com/miketeo/pysmb/blob/master/python2/smb/SMBConnection.py#L126 and append those to the list instead...
eg: connect to each ip and check for shares on the network

**Comment by miigotu** [comment]
PS: Don't be bothered by my babble much, I too have barely been sleeping =P

**Comment by m00n22-zz** [comment]
@miigotu I'm the one requested this feature, not knowing it would be picked up so quickly by @sharkykh . As stated in my original post #3457 I'm really enjoying Sickrage but don't like the current color scheme. Asfare as popularity on this feature I cannot comment besides the fact that I'll definitely use it. But this could lead to some nice css skins for SickRage people can share.

**Comment by sharkykh** [comment]
@miigotu I started working on this and realized that it's much simpler to iterate through the settings when you set it up like a sub-section (that's what it's called by ConfigObj):

(`config.ini`)
```ini
[NetworkShares]
    [[NFS]]
        user = miigotu
        password = sickrage
        server = HostNameOrIP
        share = sharedFolderName
``` 
(`__init__`)
```python
        for share in CFG['NetworkShares']:
            NETWORK_SHARES[share] = {}
            for key in CFG['NetworkShares'][share]:
                NETWORK_SHARES[share][key] = check_setting_str(CFG['NetworkShares'], share, key)
```

then sickbeard.NETWORK_SHARES is:
```
NETWORK_SHARES = {dict}
    u'NFS' = {dict}
        u'password' = {unicode} u'sickrage'
        u'server' = {unicode} u'HostNameOrIP'
        u'share' = {unicode} u'sharedFolderName'
        u'user' = {unicode} u'miigotu'
```

**of course it will eventually store the password as encrypted**


Unfortunately , [UNC paths don't have the option to pass credentials like you suggested. which makes it more complicated](https://superuser.com/a/344933)
So I really do need to make the whole drive mapping for windows from the config to achieve this.

File: gui/slick/views/layouts/main.mako [hunk 3]
```diff
@@ -94,11 +94,15 @@
...
-        %if sickbeard.THEME_NAME != "light":
+        % if sickbeard.THEME_NAME != "light":
             <link rel="stylesheet" type="text/css" href="${srRoot}/css/${sickbeard.THEME_NAME}.css?${sbPID}" />
-        %endif
+        % endif
 
         <%block name="css" />
+
+        % if sickbeard.CUSTOM_CSS:
+            <link rel="stylesheet" type="text/css" href="${srRoot}/ui/custom.css" />
```
Comment (OmgImAlexis) [inline gui/slick/views/layouts/main.mako:104]: Why is the path hardcoded if there's a config option for `CUSTOM_CSS_PATH`?

Comment (miigotu) [reply gui/slick/views/layouts/main.mako:104]: Webserve has a view for that route, which returns the true custom css

File: sickbeard/webserve.py (hunk 2 above)
Comment (miigotu) [reply sickbeard/webserve.py:655]: @OmgImAlexis here

File: gui/slick/views/layouts/main.mako (hunk 3 above)
Comment (OmgImAlexis) [reply gui/slick/views/layouts/main.mako:104]: It'd make more sense to just serve it as a file so caching works.

Comment (miigotu) [reply gui/slick/views/layouts/main.mako:104]: Did you try it and see if caching works?  If it doesnt then a header just needs fixed in the view.  Serving it as a file would require us to copy it into the gui folder, or somewhere, and add a static file handler for it.  For such a simple feature that is unlikely to be used I dont think putting too much work into it is warranted, unless the community starts making all kinds of css for SR. Then we can put all kinds of time into it. But for now this was just one user requesting this.

Comment (sharkykh) [reply gui/slick/views/layouts/main.mako:104]: For the record, I was following the same concept as sickrage_background.

Comment (miigotu) [reply gui/slick/views/layouts/main.mako:104]: No worries man, the custom.css is THE ONLY css that is caching correctly.
image

Comment (miigotu) [reply gui/slick/views/layouts/main.mako:104]: We do need to add ?${srPID}(or whatever the pid var is) to it though so cache breaks on restart.

**Pull request merged**